<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>TaxSEA • TaxSEA</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="TaxSEA">
<meta name="description" content="Taxon Set Enrichment Analysis">
<meta property="og:description" content="Taxon Set Enrichment Analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">TaxSEA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-guides" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Guides</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-guides">
<li><a class="dropdown-item" href="articles/why-taxon-set-enrichment.html">Why use taxon set enrichment?</a></li>
    <li><a class="dropdown-item" href="articles/FAQs.html">FAQs</a></li>
    <li><a class="dropdown-item" href="articles/Taxonomic-aggregation.html">Taxonomic aggregation</a></li>
    <li><a class="dropdown-item" href="articles/TaxSEA.html">Using TaxSEA</a></li>
    <li><a class="dropdown-item" href="articles/ORA_vs_ES.html">Analysis types</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/feargalr/taxsea/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="taxsea-taxon-set-enrichment-analysis">TaxSEA: Taxon Set Enrichment Analysis<a class="anchor" aria-label="anchor" href="#taxsea-taxon-set-enrichment-analysis"></a>
</h1></div>
<p><a href="https://bioconductor.org/packages/TaxSEA" class="external-link"><img src="https://www.bioconductor.org/shields/years-in-bioc/TaxSEA.svg" alt="Bioconductor"></a><br>
Available on <a href="https://bioconductor.org/packages/TaxSEA" class="external-link"><strong>Bioconductor</strong></a><br><a href="https://academic.oup.com/bib/article/26/2/bbaf173/8116684" class="external-link">Read the paper in Briefings in Bioinformatics</a></p>
<div class="section level3">
<h3 id="what-is-taxsea">What is TaxSEA?<a class="anchor" aria-label="anchor" href="#what-is-taxsea"></a>
</h3>
<p>TaxSEA is technique to move from analysisng individual microbes to looking at groups of microbes with a shared functional characteristic.</p>
<p>Traditional microbiome analysis tools often look at individual microbes going up or down. This sounds sensible but in human microbiome studies it can fail apart because everyone’s microbiome is unique, with few species shared across people.</p>
<p>TaxSEA doesn’t test for individual species but rather asks if there a group of species with a shared characteristic that is altered between groups of people. For example are we seeing a shift in bugs with increased oxygen tolerance, or an increase in bugs which produce a certain metabolite or enzyme, or even are we seeing a change in bugs which have a known disease association in humans. If you want to know more check out our paper, but we find this approach to be far more reproducible than standard Differential Abundance (DA) analysis as well as capable of extracting biologically meaningful patterns. TaxSEA usually can run within seconds with standard hardware.</p>
<p>TaxSEA supports <strong>three complementary analysis modes</strong>:</p>
<ol style="list-style-type: decimal">
<li><strong>Enrichment using public reference databases</strong></li>
<li>
<strong>ORA – Over-Representation Analysis</strong><br>
</li>
<li><strong>Enrichment using taxonomic sets</strong></li>
</ol>
<p>Modes 1 and 2 Both use the same taxon-set databases but answer slightly different questions. Mode 1 which is the default we reccomend in most cases takes as input a list of bacteria a rank (E.g. fold change). It then tests whether any of the taxon sets in the database are skewed to one end of the distribution or another. Mode 2 uses the same database but only takes as input a list of bacterial names. While answering a similar question this approach is often less powerful as it requires a hard cut-off to select bacteria of interest.</p>
<p>Mode 3 allows users to test for differences at a particular taxonomic rank. For example instead of just summing up or aggregating all the species in a genus. We see if the distribution of species within a genus is different between groups. This is a more powerful approach as simply summing/aggregating to a rank can risk missing when you can get shifts within a taxon. This approach is implemented in the taxon_set_ranks() function.</p>
<p>In short, TaxSEA aims to make it easy to intepret changes in your microbiome data.</p>
<p>TaxSEA takes as input a vector of species names and a rank. For example log2 fold changes or Spearman’s rho.</p>
<p>Note: Although TaxSEA in principle can be applied to microbiome data from any source, the databases utilized largely cover human associated microbiomes and the human gut microbiome in particular. As such the database testing in TaxSEA will likely perform best on human gut microbiome data.</p>
</div>
<div class="section level3">
<h3 id="taxon-set-database">Taxon set database<a class="anchor" aria-label="anchor" href="#taxon-set-database"></a>
</h3>
<p>By default TaxSEA utilizes taxon sets generated from six reference databases (<strong>BacDive</strong>,<strong>gutMGene</strong>, <strong>GMrepo v2</strong>, <strong>MiMeDB</strong>, <strong>mBodyMap</strong>, <strong>BugSigDB</strong>). See below for examples of using custom databases or taxonomically defined taxon sets.</p>
<p>Please cite the appropriate database if using:</p>
<ul>
<li>Schober et al. BacDive in 2025: the core database for prokaryotic strain data. Nucleic Acids Res. 2025.</li>
<li>Cheng et al. gutMGene: a comprehensive database for target genes of gut microbes and microbial metabolites. Nucleic Acids Res. 2022.</li>
<li>Dai et al. GMrepo v2: a curated human gut microbiome database with special focus on disease markers and cross-dataset comparison Nucleic Acids Res. 2022.</li>
<li>Wishart et. al. MiMeDB: the Human Microbial Metabolome Database Nucleic Acids Res. 2023.</li>
<li>Jin et al. mBodyMap: a curated database for microbes across human body and their associations with health and diseases. Nucleic Acids Res. 2022.</li>
<li>Geistlinger et al. BugSigDB captures patterns of differential abundance across a broad range of host-associated microbial signatures. Nature Biotech. 2023.</li>
</ul>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"TaxSEA"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h3>
<div class="section level4">
<h4 id="quick-start-mode-1-enrichment-rank-based">Quick start (mode 1, enrichment rank based)<a class="anchor" aria-label="anchor" href="#quick-start-mode-1-enrichment-rank-based"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feargalr/taxsea" class="external-link">TaxSEA</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Retrieve taxon sets containing Bifidobacterium longum.</span></span>
<span><span class="va">blong.sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/get_taxon_sets.html">get_taxon_sets</a></span><span class="op">(</span>taxon<span class="op">=</span><span class="st">"Bifidobacterium_longum"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run TaxSEA with test data provided</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">TaxSEA_test_data</span><span class="op">)</span></span>
<span><span class="va">taxsea_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/TaxSEA.html">TaxSEA</a></span><span class="op">(</span>taxon_ranks<span class="op">=</span><span class="va">TaxSEA_test_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Enrichments among metabolite producers from gutMgene and MiMeDB</span></span>
<span><span class="va">metabolites.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">Metabolite_producers</span></span>
<span></span>
<span><span class="co">#Enrichments among health and disease signatures from GMRepoV2 and mBodyMap</span></span>
<span><span class="va">disease.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">Health_associations</span></span>
<span></span>
<span><span class="co">#Enrichments among published associations from BugSigDB</span></span>
<span><span class="va">bsdb.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">BugSigdB</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="quick-start-mode-2-ora">Quick start (mode 2, ORA)<a class="anchor" aria-label="anchor" href="#quick-start-mode-2-ora"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feargalr/taxsea" class="external-link">TaxSEA</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># A list of oral taxa</span></span>
<span><span class="va">test_ORA_input</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"Streptococcus_mitis"</span>,</span>
<span>  <span class="st">"Haemophilus_parainfluenzae"</span>,</span>
<span>  <span class="st">"Fusobacterium_periodonticum"</span>,</span>
<span>  <span class="st">"Fusobacterium_nucleatum"</span>,</span>
<span>  <span class="st">"Neisseria_elongata"</span>,</span>
<span>  <span class="st">"Neisseria_flavescens"</span>,</span>
<span>  <span class="st">"Streptococcus_sanguinis"</span>,</span>
<span>  <span class="st">"Streptococcus_parasanguinis"</span>,</span>
<span>  <span class="st">"Streptococcus_salivarius"</span>,</span>
<span>  <span class="st">"Capnocytophaga_sputigena"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">taxsea_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/TaxSEA.html">TaxSEA</a></span><span class="op">(</span>input_taxa<span class="op">=</span><span class="va">test_ORA_input</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Enrichments among health and disease signatures from GMRepoV2 and mBodyMap</span></span>
<span><span class="va">disease.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">Health_associations</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="input">Input<a class="anchor" aria-label="anchor" href="#input"></a>
</h4>
<p>TaxSEA supports multiple input types, depending on the analysis mode you choose:</p>
<div class="section level5">
<h5 id="enrichment-mode-rank-based-recommended">Enrichment mode (rank-based; recommended)<a class="anchor" aria-label="anchor" href="#enrichment-mode-rank-based-recommended"></a>
</h5>
<p>You provide a <strong>named numeric vector</strong> where:</p>
<ul>
<li>The <strong>names</strong> are taxa (species or genus)</li>
<li>The <strong>values</strong> are ranks or signed statistics (e.g. log2 fold changes, correlation coefficients)</li>
</ul>
<p>In enrichment mode, the input <strong>must include all taxa that were tested</strong> in your analysis.<br>
Do <strong>not</strong> apply a significance threshold or pre-filter taxa before running TaxSEA.<br>
The method relies on the <strong>full distribution</strong> of values to detect coordinated shifts across taxon sets.</p>
</div>
<div class="section level5">
<h5 id="ora-mode-set-based">ORA mode (set-based)<a class="anchor" aria-label="anchor" href="#ora-mode-set-based"></a>
</h5>
<p>You provide a <strong>character vector of taxa names</strong> representing taxa of interest (i.e. a “hit list”).</p>
<ul>
<li>No ranking or numeric values are required</li>
<li>TaxSEA tests whether these taxa are over-represented in predefined taxon sets</li>
<li>This mode is most appropriate for presence/absence or binary data</li>
</ul>
<p>Input taxa names should be in the format of like one of the following</p>
<ul>
<li>Species name. E.g. “Bifidobacterium longum”, “Bifidobacterium_longum”</li>
<li>Genus name. E.g. “Bifidobacterium”</li>
<li>NCBI ID E.g. 216816</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Input IDs with the full taxonomic lineage should be split up. E.g.</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="st">"d__Bacteria.p__Actinobacteriota.c__Actinomycetes.o__Bifidobacteriales.f__Bifidobacteriaceae.g__Bifidobacterium"</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">x</span>,split<span class="op">=</span><span class="st">"\\."</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"g__"</span>,<span class="st">""</span>,<span class="va">x</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Running this through a vector of IDs may look something like the following</span></span>
<span><span class="co">#new_ids &lt;- sapply(as.character(old_ids),function(y) {strsplit(x = y,split="\\.")[[1]][6]})</span></span>
<span><span class="co">#new_ids &lt;- gsub("g__","",new_ids)</span></span>
<span></span>
<span><span class="co">## Example test data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feargalr/taxsea" class="external-link">TaxSEA</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"TaxSEA_test_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">TaxSEA_test_data</span><span class="op">)</span>,<span class="fl">4</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="run-taxsea-with-test-data">Run TaxSEA with test data<a class="anchor" aria-label="anchor" href="#run-taxsea-with-test-data"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"TaxSEA_test_data"</span><span class="op">)</span></span>
<span><span class="va">taxsea_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/TaxSEA.html">TaxSEA</a></span><span class="op">(</span>taxon_ranks<span class="op">=</span><span class="va">TaxSEA_test_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Enrichments among metabolite producers from gutMgene and MiMeDB</span></span>
<span><span class="va">metabolites.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">Metabolite_producers</span></span>
<span></span>
<span><span class="co">#Enrichments among health and disease signatures from GMRepoV2 and mBodyMap</span></span>
<span><span class="va">disease.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">Health_associations</span></span>
<span></span>
<span><span class="co">#Enrichments among published associations from BugSigDB</span></span>
<span><span class="va">bsdb.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">BugSigDB</span></span></code></pre></div>
<div class="section level4">
<h4 id="test-data">Test data<a class="anchor" aria-label="anchor" href="#test-data"></a>
</h4>
<p>The test data provided with TaxSEA consists of log2 fold changes comparing between healthy and IBD. The count data for this was downloaded from curatedMetagenomeData and fold changes generated with LinDA.</p>
<ul>
<li>Hall et al. A novel Ruminococcus gnavus clade enriched in inflammatory bowel disease patients** Genome Med. 2017 Nov 28;9(1):103.</li>
<li>Pasolli et al. Accessible, curated metagenomic data through ExperimentHub. Nat Methods. 2017 Oct 31;14(11):1023-1024. doi: 10.1038/nmeth.4468.</li>
<li>Zhou et al. LinDA: linear models for differential abundance analysis of microbiome compositional data. Genome Biol. 2022 Apr 14;23(1):95.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">head</span>(<span class="fu">sample</span>(TaxSEA_test_data),<span class="dv">3</span>)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>Bacteroides_thetaiotaomicron           Blautia_sp_CAG_257          Ruminococcus bromii</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>                       <span class="fl">1.908</span>                        <span class="fl">3.650</span>                       <span class="sc">-</span><span class="fl">5.038</span> </span></code></pre></div>
</div>
<div class="section level4">
<h4 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h4>
<p>The output is a list of three data frames providing enrichment results for metabolite producers, health/disease associations, and published signatures from BugSigDB. Each dataframe has 5 columns</p>
<ul>
<li>taxonSetName - The name of the taxon set tested</li>
<li>median_rank - This is simply the median rank across all detected members in the set. This allows you to see the direction of change</li>
<li>P value - Kolmogorov-Smirnov test P value.</li>
<li>FDR - P value adjusted for multiple testing.</li>
<li>TaxonSet - Returns list of taxa in the set to show what is driving the signal</li>
</ul>
</div>
<div class="section level4">
<h4 id="custom-databases">Custom databases<a class="anchor" aria-label="anchor" href="#custom-databases"></a>
</h4>
<p>Many users may want to utilise TaxSEA with a custom database. For example for testing if there is a flag in the TaxSEA function “custom_db” which expects as input a named list of vectors. This is the same format as the default TaxSEA database. Note: using the custom_db flag disables the automatic ID conversion and NCBI API lookup. However we have functionality available via other functions</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform enrichment analysis using TaxSEA</span></span>
<span><span class="va">custom_taxsea_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/TaxSEA.html">TaxSEA</a></span><span class="op">(</span>taxon_ranks <span class="op">=</span> <span class="va">log2_fold_changes</span>, custom_db <span class="op">=</span> <span class="va">custom_taxon_sets</span><span class="op">)</span></span>
<span><span class="va">custom_taxsea_results</span> <span class="op">&lt;-</span> <span class="va">custom_taxsea_results</span><span class="op">$</span><span class="va">custom_sets</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="testing-for-differences-in-taxonomically-defined-sets">Testing for differences in taxonomically defined sets<a class="anchor" aria-label="anchor" href="#testing-for-differences-in-taxonomically-defined-sets"></a>
</h4>
<p>In addition to taxon sets defined from public databases by, users can define sets based on taxonomy from their own data. We have implemtned this in the taxon_rank_sets() function which takes as input an set of taxa and ranks (similar to enrichment testing above) and then either a table of taxonomic lineage or a TreeSummarizedExperiment object. See below for examples with both approaches.</p>
<p>Current methods to test at higher taxonomic levels (e.g., phylum or family), involve aggregating counts. However, this approach may miss opposing shifts in individual species which cancel each other out, obscuring meaningful biological patterns. For instance, antibiotic treatment may suppress certain species while allowing resistant species within the same genus to expand and occupy the vacant niche, creating an ecological shift that appears as no net change at broader taxonomic levels. Whereas the database testing works best with shotgun metagenomic data in order to have sufficient resolution. This approach is much more widely applicable including to testing with amplicon data such as from 16S rRNA gene sequencing.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#### Enrichment testing with taxonomically defined ranks  </span></span>
<span><span class="co"># --- Example 1: Data frame input ---</span></span>
<span><span class="co"># Create a lineage data frame (e.g., parsed from curatedMetagenomicData)</span></span>
<span><span class="co"># The 'species' column must match the names in taxon_ranks.</span></span>
<span></span>
<span><span class="va">lineage_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cutibacterium_acnes"</span>, <span class="st">"Klebsiella_pneumoniae"</span>,</span>
<span>              <span class="st">"Propionibacterium_humerusii"</span>, <span class="st">"Moraxella_osloensis"</span>,</span>
<span>              <span class="st">"Enhydrobacter_aerosaccus"</span>, <span class="st">"Staphylococcus_capitis"</span>,</span>
<span>              <span class="st">"Staphylococcus_epidermidis"</span>, <span class="st">"Staphylococcus_aureus"</span>,</span>
<span>              <span class="st">"Escherichia_coli"</span>, <span class="st">"Enterobacter_cloacae"</span>,</span>
<span>              <span class="st">"Pseudomonas_aeruginosa"</span>, <span class="st">"Acinetobacter_baumannii"</span>,</span>
<span>              <span class="st">"Lactobacillus_rhamnosus"</span>, <span class="st">"Lactobacillus_acidophilus"</span>,</span>
<span>              <span class="st">"Bifidobacterium_longum"</span>, <span class="st">"Bifidobacterium_breve"</span><span class="op">)</span>,</span>
<span>  kingdom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Bacteria"</span>, <span class="fl">16</span><span class="op">)</span>,</span>
<span>  phylum <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Actinobacteria"</span>, <span class="st">"Proteobacteria"</span>,</span>
<span>             <span class="st">"Actinobacteria"</span>, <span class="st">"Proteobacteria"</span>,</span>
<span>             <span class="st">"Proteobacteria"</span>, <span class="st">"Firmicutes"</span>,</span>
<span>             <span class="st">"Firmicutes"</span>, <span class="st">"Firmicutes"</span>,</span>
<span>             <span class="st">"Proteobacteria"</span>, <span class="st">"Proteobacteria"</span>,</span>
<span>             <span class="st">"Proteobacteria"</span>, <span class="st">"Proteobacteria"</span>,</span>
<span>             <span class="st">"Firmicutes"</span>, <span class="st">"Firmicutes"</span>,</span>
<span>             <span class="st">"Actinobacteria"</span>, <span class="st">"Actinobacteria"</span><span class="op">)</span>,</span>
<span>  class <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Actinobacteria"</span>, <span class="st">"Gammaproteobacteria"</span>,</span>
<span>            <span class="st">"Actinobacteria"</span>, <span class="st">"Gammaproteobacteria"</span>,</span>
<span>            <span class="st">"Alphaproteobacteria"</span>, <span class="st">"Bacilli"</span>,</span>
<span>            <span class="st">"Bacilli"</span>, <span class="st">"Bacilli"</span>,</span>
<span>            <span class="st">"Gammaproteobacteria"</span>, <span class="st">"Gammaproteobacteria"</span>,</span>
<span>            <span class="st">"Gammaproteobacteria"</span>, <span class="st">"Gammaproteobacteria"</span>,</span>
<span>            <span class="st">"Bacilli"</span>, <span class="st">"Bacilli"</span>,</span>
<span>            <span class="st">"Actinobacteria"</span>, <span class="st">"Actinobacteria"</span><span class="op">)</span>,</span>
<span>  order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Propionibacteriales"</span>, <span class="st">"Enterobacterales"</span>,</span>
<span>            <span class="st">"Propionibacteriales"</span>, <span class="st">"Pseudomonadales"</span>,</span>
<span>            <span class="st">"Rhodospirillales"</span>, <span class="st">"Bacillales"</span>,</span>
<span>            <span class="st">"Bacillales"</span>, <span class="st">"Bacillales"</span>,</span>
<span>            <span class="st">"Enterobacterales"</span>, <span class="st">"Enterobacterales"</span>,</span>
<span>            <span class="st">"Pseudomonadales"</span>, <span class="st">"Pseudomonadales"</span>,</span>
<span>            <span class="st">"Lactobacillales"</span>, <span class="st">"Lactobacillales"</span>,</span>
<span>            <span class="st">"Bifidobacteriales"</span>, <span class="st">"Bifidobacteriales"</span><span class="op">)</span>,</span>
<span>  family <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Propionibacteriaceae"</span>, <span class="st">"Enterobacteriaceae"</span>,</span>
<span>             <span class="st">"Propionibacteriaceae"</span>, <span class="st">"Moraxellaceae"</span>,</span>
<span>             <span class="st">"Rhodospirillaceae"</span>, <span class="st">"Staphylococcaceae"</span>,</span>
<span>             <span class="st">"Staphylococcaceae"</span>, <span class="st">"Staphylococcaceae"</span>,</span>
<span>             <span class="st">"Enterobacteriaceae"</span>, <span class="st">"Enterobacteriaceae"</span>,</span>
<span>             <span class="st">"Pseudomonadaceae"</span>, <span class="st">"Moraxellaceae"</span>,</span>
<span>             <span class="st">"Lactobacillaceae"</span>, <span class="st">"Lactobacillaceae"</span>,</span>
<span>             <span class="st">"Bifidobacteriaceae"</span>, <span class="st">"Bifidobacteriaceae"</span><span class="op">)</span>,</span>
<span>  genus <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cutibacterium"</span>, <span class="st">"Klebsiella"</span>,</span>
<span>            <span class="st">"Cutibacterium"</span>, <span class="st">"Moraxella"</span>,</span>
<span>            <span class="st">"Enhydrobacter"</span>, <span class="st">"Staphylococcus"</span>,</span>
<span>            <span class="st">"Staphylococcus"</span>, <span class="st">"Staphylococcus"</span>,</span>
<span>            <span class="st">"Escherichia"</span>, <span class="st">"Enterobacter"</span>,</span>
<span>            <span class="st">"Pseudomonas"</span>, <span class="st">"Acinetobacter"</span>,</span>
<span>            <span class="st">"Lactobacillus"</span>, <span class="st">"Lactobacillus"</span>,</span>
<span>            <span class="st">"Bifidobacterium"</span>, <span class="st">"Bifidobacterium"</span><span class="op">)</span>,</span>
<span>  stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">fc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">16</span><span class="op">)</span>, <span class="va">lineage_df</span><span class="op">$</span><span class="va">species</span><span class="op">)</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/taxon_rank_sets.html">taxon_rank_sets</a></span><span class="op">(</span><span class="va">fc</span>, <span class="va">lineage_df</span>, min_set_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span>
<span><span class="va">results</span><span class="op">$</span><span class="va">family</span></span>
<span></span>
<span></span>
<span><span class="co"># --- Example 2: SummarizedExperiment / TreeSummarizedExperiment input ---</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://microbiome.github.io/mia/" class="external-link">mia</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feargalr/taxsea" class="external-link">TaxSEA</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ALDEx2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">GlobalPatterns</span>, package <span class="op">=</span> <span class="st">"mia"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">GlobalPatterns</span></span>
<span></span>
<span><span class="co"># Subset to two sample types for a two-group comparison</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">tse</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">$</span><span class="va">SampleType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Feces"</span>, <span class="st">"Soil"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Filter low-prevalence features (present in at least 2 samples)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">tse</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"counts"</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">2</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Run ALDEx2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">aldex.clr</span><span class="op">(</span></span>
<span>  reads <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">tse</span>, <span class="st">"counts"</span><span class="op">)</span>,</span>
<span>  conds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">$</span><span class="va">SampleType</span><span class="op">)</span>,</span>
<span>  mc.samples <span class="op">=</span> <span class="fl">128</span>,</span>
<span>  denom <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">x_tt</span> <span class="op">&lt;-</span> <span class="fu">aldex.ttest</span><span class="op">(</span><span class="va">x</span>, paired.test <span class="op">=</span> <span class="cn">FALSE</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">x_effect</span> <span class="op">&lt;-</span> <span class="fu">aldex.effect</span><span class="op">(</span><span class="va">x</span>, CI <span class="op">=</span> <span class="cn">TRUE</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">aldex_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">x_tt</span>, <span class="va">x_effect</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Use ALDEx2 effect size as ranking metric</span></span>
<span><span class="va">fc</span> <span class="op">&lt;-</span> <span class="va">aldex_out</span><span class="op">$</span><span class="va">effect</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">aldex_out</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run taxon rank set enrichment directly from the TSE</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/taxon_rank_sets.html">taxon_rank_sets</a></span><span class="op">(</span><span class="va">fc</span>, <span class="va">tse</span>, min_set_size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span>    <span class="co"># Kingdom, Phylum, Class, Order, Family, Genus</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">Family</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">Genus</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="visualisation-of-taxsea-output">Visualisation of TaxSEA output.<a class="anchor" aria-label="anchor" href="#visualisation-of-taxsea-output"></a>
</h4>
<p><img width="956" alt="TaxSEA visualisation" src="reference/figures/taxsea_viz1.png"></p>
<p>The results above were generated by using TaxSEA on the output of a differential abundance analysis comparing between disease and control. The input was per species log2 fold changes between taxa in Inflammatory Bowel disease and control. TaxSEA identified a significant depletion in the producers of certain short chain fatty acids. Using barplots we can show the overall signatures identified as significantly different. We can then highlight the individual species contributing to this signature on a volcano plot.</p>
</div>
<div class="section level4">
<h4 id="bugsigdb">BugSigDB<a class="anchor" aria-label="anchor" href="#bugsigdb"></a>
</h4>
<p>The format of BugSigDB is that each publication is entered as a “Study”, and within this there is different experiments and signatures. For example one of the signatures may be taxa increased in an experiment, and another signature is taxa that are decreased. Users can find out more by querying the BugSigDB. In recent updates BugSigDB has moved to using the PubMed ID as the study ID (Although not all studies have these yet). See below for an example.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/waldronlab/bugsigdbr" class="external-link">bugsigdbr</a></span><span class="op">)</span> <span class="co">#This package is installable via Bioconductor</span></span>
<span><span class="va">bsdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bugsigdbr/man/importBugSigDB.html" class="external-link">importBugSigDB</a></span><span class="op">(</span><span class="op">)</span> <span class="co">#Import database </span></span>
<span></span>
<span><span class="co">#E.g. if the BugSigDB identifier you found enriched was bsdb:11/1/1</span></span>
<span><span class="co">#This is Study 11, Experiment 1, Signature 1</span></span>
<span><span class="va">bsdb</span><span class="op">[</span><span class="va">bsdb</span><span class="op">$</span><span class="va">Study</span><span class="op">==</span><span class="st">"Study 11"</span> <span class="op">&amp;</span> </span>
<span>     <span class="va">bsdb</span><span class="op">$</span><span class="va">Experiment</span><span class="op">==</span><span class="st">"Experiment 1"</span> <span class="op">&amp;</span> </span>
<span>     <span class="va">bsdb</span><span class="op">$</span><span class="va">Signature</span><span class="op">==</span><span class="st">"Signature 1"</span>,<span class="op">]</span></span>
<span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="taxsea-database-with-other-enrichment-tools">TaxSEA database with other enrichment tools<a class="anchor" aria-label="anchor" href="#taxsea-database-with-other-enrichment-tools"></a>
</h4>
<p>The TaxSEA function by default uses the Kolmogorov Smirnov test and the original idea was inspired by gene set enrichment analyses from RNASeq. Should users wish to use an alternative gene set enrichment analysis tool the database is formatted in such a way that should be possible. See below for an example with fast gene set enrichment analysis (fgsea).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ctlab/fgsea/" class="external-link">fgsea</a></span><span class="op">)</span> <span class="co">#This package is installable via Bioconductor</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">TaxSEA_DB</span><span class="op">)</span></span>
<span><span class="co">#Convert input names to NCBI taxon ids</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">TaxSEA_test_data</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="reference/get_ncbi_taxon_ids.html">get_ncbi_taxon_ids</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">TaxSEA_test_data</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">TaxSEA_test_data</span> <span class="op">=</span> <span class="va">TaxSEA_test_data</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">TaxSEA_test_data</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#Run fgsea</span></span>
<span><span class="va">fgsea_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/fgsea/man/fgsea.html" class="external-link">fgsea</a></span><span class="op">(</span><span class="va">TaxSEA_db</span>, <span class="va">TaxSEA_test_data</span>, minSize<span class="op">=</span><span class="fl">5</span>, maxSize<span class="op">=</span><span class="fl">500</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="">
<h2 data-toc-skip></h2>
<ul class="list-unstyled">
<li><div style="text-align:center; margin-bottom: 1rem;">
<p><img src="reference/figures/hex.png" width="200" alt="TaxSEA logo"></p>
</div></li>
</ul>
</div>

<div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/TaxSEA" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/feargalr/taxsea/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/feargalr/taxsea/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing TaxSEA</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Feargal Ryan <br><small class="roles"> Author, maintainer, funder </small> <a href="https://orcid.org/0000-0002-1565-4598" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Feargal Ryan.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
