<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>TaxSEA • TaxSEA</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="TaxSEA">
<meta name="description" content="Taxon Set Enrichment Analysis">
<meta property="og:description" content="Taxon Set Enrichment Analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">TaxSEA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/feargalr/taxsea/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="taxsea-taxon-set-enrichment-analysis">TaxSEA: Taxon Set Enrichment Analysis<a class="anchor" aria-label="anchor" href="#taxsea-taxon-set-enrichment-analysis"></a>
</h1></div>
<p><a href="https://bioconductor.org/packages/TaxSEA" class="external-link"><img src="https://www.bioconductor.org/shields/years-in-bioc/TaxSEA.svg" alt="Bioconductor"></a><br>
Available on <a href="https://bioconductor.org/packages/TaxSEA" class="external-link"><strong>Bioconductor</strong></a><br><a href="https://academic.oup.com/bib/article/26/2/bbaf173/8116684" class="external-link">Read the paper in Briefings in Bioinformatics</a> <img align="right" src="reference/figures/hex.png" alt="TaxSEA" height="200"></p>
<div class="section level3">
<h3 id="what-is-taxsea">What is TaxSEA?<a class="anchor" aria-label="anchor" href="#what-is-taxsea"></a>
</h3>
<p>Traditional differential abundance (DA) analysis looks at individual microbes going up or down. This sounds sensible but in human microbiome studies it often falls apart because everyone’s microbiome is unique, with few species shared across people. TaxSEA doesn’t seem to test for individual species are different but rather is there a group of species with a shared characteristic that is different. For example are we seeing a shift in bugs with increased oxygen tolerance, or an increase in bugs which produce a certain metabolite or enzyme, or even are we seeing a change in bugs which have a known disease association in humans. If you want to know more check out our paper, but we find this approach to be far more reproducible than standard DA analysis as well as capable of extracting biologically meaningful patterns.</p>
<p>TaxSEA can also test for differences at a particular taxonomic rank. For example instead of just summing up all the species in a genus. We see if the distribution of species within a genus is different between groups.</p>
<p>In short, TaxSEA makes it easy to intepret changes in your microbiome data.</p>
<p>TaxSEA takes as input a vector of genus or species names and a rank. For example log2 fold changes or Spearman’s rho. TaxSEA then uses a Kolmogorov-Smirnov test to identify if a particular group of species or genera (i.e. a set of taxa such as butyrate producers) are skewed to one end of the distribution.</p>
<p>Note: Although TaxSEA in principle can be applied to microbiome data from any source, the databases utilized largely cover human associated microbiomes and the human gut microbiome in particular. As such TaxSEA will likely perform best on human gut microbiome data.</p>
</div>
<div class="section level3">
<h3 id="taxon-set-database">Taxon set database<a class="anchor" aria-label="anchor" href="#taxon-set-database"></a>
</h3>
<p>By default TaxSEA utilizes taxon sets generated from six reference databases (<strong>BacDive</strong>,<strong>gutMGene</strong>, <strong>GMrepo v2</strong>, <strong>MiMeDB</strong>, <strong>mBodyMap</strong>, <strong>BugSigDB</strong>). See below for examples of using custom databases or taxonomically defined taxon sets.</p>
<p>Please cite the appropriate database if using:</p>
<ul>
<li>Schober et al. BacDive in 2025: the core database for prokaryotic strain data. Nucleic Acids Res. 2025.</li>
<li>Cheng et al. gutMGene: a comprehensive database for target genes of gut microbes and microbial metabolites. Nucleic Acids Res. 2022.</li>
<li>Dai et al. GMrepo v2: a curated human gut microbiome database with special focus on disease markers and cross-dataset comparison Nucleic Acids Res. 2022.</li>
<li>Wishart et. al. MiMeDB: the Human Microbial Metabolome Database Nucleic Acids Res. 2023.</li>
<li>Jin et al. mBodyMap: a curated database for microbes across human body and their associations with health and diseases. Nucleic Acids Res. 2022.</li>
<li>Geistlinger et al. BugSigDB captures patterns of differential abundance across a broad range of host-associated microbial signatures. Nature Biotech. 2023.</li>
</ul>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"TaxSEA"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h3>
<div class="section level4">
<h4 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feargalr/taxsea" class="external-link">TaxSEA</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Retrieve taxon sets containing Bifidobacterium longum.</span></span>
<span><span class="va">blong.sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/get_taxon_sets.html">get_taxon_sets</a></span><span class="op">(</span>taxon<span class="op">=</span><span class="st">"Bifidobacterium_longum"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run TaxSEA with test data provided</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">TaxSEA_test_data</span><span class="op">)</span></span>
<span><span class="va">taxsea_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/TaxSEA.html">TaxSEA</a></span><span class="op">(</span>taxon_ranks<span class="op">=</span><span class="va">TaxSEA_test_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Enrichments among metabolite producers from gutMgene and MiMeDB</span></span>
<span><span class="va">metabolites.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">Metabolite_producers</span></span>
<span></span>
<span><span class="co">#Enrichments among health and disease signatures from GMRepoV2 and mBodyMap</span></span>
<span><span class="va">disease.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">Health_associations</span></span>
<span></span>
<span><span class="co">#Enrichments among published associations from BugSigDB</span></span>
<span><span class="va">bsdb.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">BugSigdB</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="input">Input<a class="anchor" aria-label="anchor" href="#input"></a>
</h4>
<p>All that is required for TaxSEA is a vector in R containing ranks (e.g. log2 fold changes) and names (E.g. species/genus). TaxSEA will not work for ranks higher than species or genus. The input should be for all taxa tested, and not limited to only a pre-defined set (e.g. do not use a threshold for significance or remove any taxa). See example below for format. TaxSEA will lookup and convert taxon names to NCBI taxonomic identifiers. TaxSEA stores a commonly observed identifiers internally and so will only look up whatever is not covered to save time.</p>
<p>Input IDs should be in the format of like one of the following</p>
<ul>
<li>Species name. E.g. “Bifidobacterium longum”, “Bifidobacterium_longum”</li>
<li>Genus name. E.g. “Bifidobacterium”</li>
<li>NCBI ID E.g. 216816</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Input IDs with the full taxonomic lineage should be split up. E.g.</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="st">"d__Bacteria.p__Actinobacteriota.c__Actinomycetes.o__Bifidobacteriales.f__Bifidobacteriaceae.g__Bifidobacterium"</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">x</span>,split<span class="op">=</span><span class="st">"\\."</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"g__"</span>,<span class="st">""</span>,<span class="va">x</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Running this through a vector of IDs may look something like the following</span></span>
<span><span class="co">#new_ids &lt;- sapply(as.character(old_ids),function(y) {strsplit(x = y,split="\\.")[[1]][6]})</span></span>
<span><span class="co">#new_ids &lt;- gsub("g__","",new_ids)</span></span>
<span></span>
<span><span class="co">## Example test data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feargalr/taxsea" class="external-link">TaxSEA</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"TaxSEA_test_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">TaxSEA_test_data</span><span class="op">)</span>,<span class="fl">4</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="run-taxsea-with-test-data">Run TaxSEA with test data<a class="anchor" aria-label="anchor" href="#run-taxsea-with-test-data"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"TaxSEA_test_data"</span><span class="op">)</span></span>
<span><span class="va">taxsea_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/TaxSEA.html">TaxSEA</a></span><span class="op">(</span>taxon_ranks<span class="op">=</span><span class="va">TaxSEA_test_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Enrichments among metabolite producers from gutMgene and MiMeDB</span></span>
<span><span class="va">metabolites.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">Metabolite_producers</span></span>
<span></span>
<span><span class="co">#Enrichments among health and disease signatures from GMRepoV2 and mBodyMap</span></span>
<span><span class="va">disease.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">Health_associations</span></span>
<span></span>
<span><span class="co">#Enrichments among published associations from BugSigDB</span></span>
<span><span class="va">bsdb.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">BugSigDB</span></span></code></pre></div>
<div class="section level4">
<h4 id="test-data">Test data<a class="anchor" aria-label="anchor" href="#test-data"></a>
</h4>
<p>The test data provided with TaxSEA consists of log2 fold changes comparing between healthy and IBD. The count data for this was downloaded from curatedMetagenomeData and fold changes generated with LinDA.</p>
<ul>
<li>Hall et al. A novel Ruminococcus gnavus clade enriched in inflammatory bowel disease patients** Genome Med. 2017 Nov 28;9(1):103.</li>
<li>Pasolli et al. Accessible, curated metagenomic data through ExperimentHub. Nat Methods. 2017 Oct 31;14(11):1023-1024. doi: 10.1038/nmeth.4468.</li>
<li>Zhou et al. LinDA: linear models for differential abundance analysis of microbiome compositional data. Genome Biol. 2022 Apr 14;23(1):95.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">head</span>(<span class="fu">sample</span>(TaxSEA_test_data),<span class="dv">3</span>)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>Bacteroides_thetaiotaomicron           Blautia_sp_CAG_257          Ruminococcus bromii</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>                       <span class="fl">1.908</span>                        <span class="fl">3.650</span>                       <span class="sc">-</span><span class="fl">5.038</span> </span></code></pre></div>
</div>
<div class="section level4">
<h4 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h4>
<p>The output is a list of three data frames providing enrichment results for metabolite producers, health/disease associations, and published signatures from BugSigDB. Each dataframe has 5 columns</p>
<ul>
<li>taxonSetName - The name of the taxon set tested</li>
<li>median_rank - This is simply the median rank across all detected members in the set. This allows you to see the direction of change</li>
<li>P value - Kolmogorov-Smirnov test P value.</li>
<li>FDR - P value adjusted for multiple testing.</li>
<li>TaxonSet - Returns list of taxa in the set to show what is driving the signal</li>
</ul>
</div>
<div class="section level4">
<h4 id="custom-databases">Custom databases<a class="anchor" aria-label="anchor" href="#custom-databases"></a>
</h4>
<p>Many users may want to utilise TaxSEA with a custom database. For example for testing if there is a flag in the TaxSEA function “custom_db” which expects as input a named list of vectors. This is the same format as the default TaxSEA database. Note: using the custom_db flag disables the automatic ID conversion and NCBI API lookup. However we have functionality available via other functions</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform enrichment analysis using TaxSEA</span></span>
<span><span class="va">custom_taxsea_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/TaxSEA.html">TaxSEA</a></span><span class="op">(</span>taxon_ranks <span class="op">=</span> <span class="va">log2_fold_changes</span>, custom_db <span class="op">=</span> <span class="va">custom_taxon_sets</span><span class="op">)</span></span>
<span><span class="va">custom_taxsea_results</span> <span class="op">&lt;-</span> <span class="va">custom_taxsea_results</span><span class="op">$</span><span class="va">custom_sets</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="testing-for-differences-in-taxonomically-defined-sets">Testing for differences in taxonomically defined sets<a class="anchor" aria-label="anchor" href="#testing-for-differences-in-taxonomically-defined-sets"></a>
</h4>
<p>In addition to taxon sets defined by function or phenotype, users can define sets based on taxonomy. Current methods to test at higher taxonomic levels (e.g., genus or family), involve aggregating counts but with this approach opposing shifts in individual species may cancel each other out, obscuring meaningful biological patterns. For instance, antibiotic treatment may suppress certain species while allowing resistant species within the same genus to expand and occupy the vacant niche, creating an ecological shift that appears as no net change at broader taxonomic levels. Here we utilise data from Chng et al. demonstrating this in a comparison between Atopic dermatitis and controls.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#### Applying TaxSEA functionality to taxonomic ranks  </span></span>
<span><span class="co"># This script applies TaxSEA to identify taxonomic enrichment at different taxonomic levels.</span></span>
<span><span class="co"># Specifically, we analyze enrichment at the family level using metagenomic data</span></span>
<span></span>
<span><span class="co"># Load required libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feargalr/taxsea" class="external-link">TaxSEA</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">curatedMetagenomicData</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">phyloseq</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">MicrobiomeStat</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load sample metadata</span></span>
<span><span class="va">metadata_all</span> <span class="op">&lt;-</span> <span class="va">sampleMetadata</span></span>
<span></span>
<span><span class="co"># Filter metadata for the specific study (ChngKR_2016)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="va">metadata_all</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">study_name</span> <span class="op">==</span> <span class="st">"ChngKR_2016"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">column_to_rownames</span><span class="op">(</span><span class="st">'sample_id'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract count data using curatedMetagenomicData</span></span>
<span><span class="va">cmd_data</span> <span class="op">&lt;-</span> <span class="fu">curatedMetagenomicData</span><span class="op">(</span></span>
<span>  pattern <span class="op">=</span> <span class="st">"ChngKR_2016.relative_abundance"</span>,</span>
<span>  counts <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  dryrun <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert the extracted data to a count matrix</span></span>
<span><span class="va">counts_data</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">cmd_data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">counts_data</span> <span class="op">&lt;-</span> <span class="va">counts_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span><span class="op">]</span>  <span class="co"># Subset to relevant samples</span></span>
<span></span>
<span><span class="co"># Filter taxa with at least one sample having counts &gt; 100</span></span>
<span><span class="va">counts_data</span> <span class="op">&lt;-</span> <span class="va">counts_data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">counts_data</span> <span class="op">&gt;</span> <span class="fl">100</span>, <span class="fl">1</span>, <span class="va">sum</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Extract species names from taxonomic strings</span></span>
<span><span class="va">species_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"s__"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts_data</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">y</span>, <span class="st">"\\|"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">7</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">species_names</span></span>
<span></span>
<span><span class="co"># Create a taxonomic lineage dataframe</span></span>
<span><span class="co"># Remove taxonomic prefixes (k__, p__, c__, etc.) and separate into taxonomic ranks</span></span>
<span></span>
<span><span class="co"># make data frame of taxon lineages</span></span>
<span><span class="va">taxon_lineages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Name <span class="op">=</span> <span class="va">species_names</span>,</span>
<span>                             Lineage <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">species_names</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Lineage <span class="op">=</span> <span class="fu">str_remove_all</span><span class="op">(</span><span class="va">Lineage</span>, <span class="st">'[kpcofgs]__'</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">separate</span><span class="op">(</span>col <span class="op">=</span> <span class="va">Lineage</span>, into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'kingdom'</span>, <span class="st">'phylum'</span>, <span class="st">'class'</span>, </span>
<span>                                   <span class="st">'order'</span>, <span class="st">'family'</span>, <span class="st">'genus'</span>, <span class="st">'species'</span><span class="op">)</span>, </span>
<span>           sep <span class="op">=</span> <span class="st">'\\|'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>name <span class="op">=</span> <span class="va">Name</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">remove_rownames</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">column_to_rownames</span><span class="op">(</span><span class="st">'name'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform differential abundance testing using LinDA</span></span>
<span><span class="va">metadata</span><span class="op">$</span><span class="va">study_condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">study_condition</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"AD"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">linda_results</span> <span class="op">&lt;-</span> <span class="fu">linda</span><span class="op">(</span></span>
<span>  feature.dat <span class="op">=</span> <span class="va">counts_data</span>,</span>
<span>  meta.dat <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>  formula <span class="op">=</span> <span class="st">'~study_condition'</span>,</span>
<span>  feature.dat.type <span class="op">=</span> <span class="st">'count'</span>,</span>
<span>  prev.filter <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract log2 fold change values for differential taxa</span></span>
<span><span class="va">linda_results</span> <span class="op">&lt;-</span> <span class="va">linda_results</span><span class="op">$</span><span class="va">output</span><span class="op">$</span><span class="va">study_conditionAD</span></span>
<span><span class="va">log2_fold_changes</span> <span class="op">&lt;-</span> <span class="va">linda_results</span><span class="op">$</span><span class="va">log2FoldChange</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">log2_fold_changes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">linda_results</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define the taxonomic rank for enrichment analysis</span></span>
<span><span class="va">selected_taxon_level</span> <span class="op">&lt;-</span> <span class="st">'genus'</span>  <span class="co"># Modify as needed (e.g., genus, phylum)</span></span>
<span></span>
<span><span class="co"># Create a named list of species grouped by taxonomic rank</span></span>
<span><span class="va">custom_taxon_sets</span> <span class="op">&lt;-</span> <span class="va">taxon_lineages</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">.data</span><span class="op">[[</span><span class="va">selected_taxon_level</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">species</span><span class="op">)</span>, .groups <span class="op">=</span> <span class="st">"drop"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">deframe</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform enrichment analysis using TaxSEA</span></span>
<span><span class="va">custom_taxsea_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/TaxSEA.html">TaxSEA</a></span><span class="op">(</span>taxon_ranks <span class="op">=</span> <span class="va">log2_fold_changes</span>, custom_db <span class="op">=</span> <span class="va">custom_taxon_sets</span><span class="op">)</span></span>
<span><span class="va">custom_taxsea_results</span> <span class="op">&lt;-</span> <span class="va">custom_taxsea_results</span><span class="op">$</span><span class="va">custom_sets</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="visualisation-of-taxsea-output">Visualisation of TaxSEA output.<a class="anchor" aria-label="anchor" href="#visualisation-of-taxsea-output"></a>
</h4>
<p><img width="956" alt="TaxSEA visualisation" src="reference/figures/taxsea_viz1.png"></p>
<p>The results above were generated by using TaxSEA on the output of a differential abundance analysis comparing between disease and control. The input was per species log2 fold changes between taxa in Inflammatory Bowel disease and control. TaxSEA identified a significant depletion in the producers of certain short chain fatty acids. Using barplots we can show the overall signatures identified as significantly different. We can then highlight the individual species contributing to this signature on a volcano plot.</p>
</div>
<div class="section level4">
<h4 id="bugsigdb">BugSigDB<a class="anchor" aria-label="anchor" href="#bugsigdb"></a>
</h4>
<p>The format of BugSigDB is that each publication is entered as a “Study”, and within this there is different experiments and signatures. For example one of the signatures may be taxa increased in an experiment, and another signature is taxa that are decreased. Users can find out more by querying the BugSigDB. In recent updates BugSigDB has moved to using the PubMed ID as the study ID (Although not all studies have these yet). See below for an example.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/waldronlab/bugsigdbr" class="external-link">bugsigdbr</a></span><span class="op">)</span> <span class="co">#This package is installable via Bioconductor</span></span>
<span><span class="va">bsdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bugsigdbr/man/importBugSigDB.html" class="external-link">importBugSigDB</a></span><span class="op">(</span><span class="op">)</span> <span class="co">#Import database </span></span>
<span></span>
<span><span class="co">#E.g. if the BugSigDB identifier you found enriched was bsdb:11/1/1</span></span>
<span><span class="co">#This is Study 11, Experiment 1, Signature 1</span></span>
<span><span class="va">bsdb</span><span class="op">[</span><span class="va">bsdb</span><span class="op">$</span><span class="va">Study</span><span class="op">==</span><span class="st">"Study 11"</span> <span class="op">&amp;</span> </span>
<span>     <span class="va">bsdb</span><span class="op">$</span><span class="va">Experiment</span><span class="op">==</span><span class="st">"Experiment 1"</span> <span class="op">&amp;</span> </span>
<span>     <span class="va">bsdb</span><span class="op">$</span><span class="va">Signature</span><span class="op">==</span><span class="st">"Signature 1"</span>,<span class="op">]</span></span>
<span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="taxsea-database-with-other-enrichment-tools">TaxSEA database with other enrichment tools<a class="anchor" aria-label="anchor" href="#taxsea-database-with-other-enrichment-tools"></a>
</h4>
<p>The TaxSEA function by default uses the Kolmogorov Smirnov test and the original idea was inspired by gene set enrichment analyses from RNASeq. Should users wish to use an alternative gene set enrichment analysis tool the database is formatted in such a way that should be possible. See below for an example with fast gene set enrichment analysis (fgsea).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ctlab/fgsea/" class="external-link">fgsea</a></span><span class="op">)</span> <span class="co">#This package is installable via Bioconductor</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">TaxSEA_DB</span><span class="op">)</span></span>
<span><span class="co">#Convert input names to NCBI taxon ids</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">TaxSEA_test_data</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="reference/get_ncbi_taxon_ids.html">get_ncbi_taxon_ids</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">TaxSEA_test_data</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">TaxSEA_test_data</span> <span class="op">=</span> <span class="va">TaxSEA_test_data</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">TaxSEA_test_data</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#Run fgsea</span></span>
<span><span class="va">fgsea_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/fgsea/man/fgsea.html" class="external-link">fgsea</a></span><span class="op">(</span><span class="va">TaxSEA_db</span>, <span class="va">TaxSEA_test_data</span>, minSize<span class="op">=</span><span class="fl">5</span>, maxSize<span class="op">=</span><span class="fl">500</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/TaxSEA" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/feargalr/taxsea/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/feargalr/taxsea/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing TaxSEA</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Feargal Ryan <br><small class="roles"> Author, maintainer, funder </small> <a href="https://orcid.org/0000-0002-1565-4598" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Feargal Ryan.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
