<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Testing higher taxonomic levels without collapsing your data • TaxSEA</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Testing higher taxonomic levels without collapsing your data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TaxSEA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-guides" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Guides</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-guides">
<li><a class="dropdown-item" href="../articles/why-taxon-set-enrichment.html">Why use taxon set enrichment?</a></li>
    <li><a class="dropdown-item" href="../articles/FAQs.html">FAQs</a></li>
    <li><a class="dropdown-item" href="../articles/Taxonomic-aggregation.html">Taxonomic aggregation</a></li>
    <li><a class="dropdown-item" href="../articles/TaxSEA.html">Using TaxSEA</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/feargalr/taxsea/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Testing higher taxonomic levels without collapsing your data</h1>
                        <h4 data-toc-skip class="author">Feargal
Ryan</h4>
            
            <h4 data-toc-skip class="date">2026-01-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/feargalr/taxsea/blob/main/vignettes/Taxonomic-aggregation.Rmd" class="external-link"><code>vignettes/Taxonomic-aggregation.Rmd</code></a></small>
      <div class="d-none name"><code>Taxonomic-aggregation.Rmd</code></div>
    </div>

    
    
<p>A common way to test for differences at higher taxonomic levels
(e.g. genus, family, order) is to aggregate counts and then run
differential abundance on the collapsed table. That approach is simple,
but it has a big flaw: opposing shifts cancel each other out. If one
species in a genus goes up and another goes down, aggregation can make
it look like nothing changed, even though the biology clearly did.</p>
<p><strong>TaxSEA offers an alternative.</strong></p>
<p>Instead of collapsing your data, you: - Run your usual analysis at
the species level - Keep the full ranked list (e.g. log2FC,
correlations, model coefficients) - Define taxonomic sets (e.g. all
species belonging to the same family) - Test whether those sets are
coordinately shifted using enrichment</p>
<p>In other words, you test higher-level taxonomy without throwing away
species-level resolution.</p>
<p>This lets you ask questions like: - Are members of this family
consistently depleted, even if no single species is significant? - Is
this genus showing a directional shift driven by many small effects? -
Are higher-level patterns being masked by species-level
heterogeneity?</p>
<p>This is particularly powerful in: - antibiotic or perturbation
experiments - situations where ecological turnover happens within
taxonomic groups</p>
<p>Below is an example performing this in R. In the future we will
incorporate this as a core function in TaxSEA</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#### Applying TaxSEA functionality to taxonomic ranks  </span></span>
<span><span class="co"># This script applies TaxSEA to identify taxonomic enrichment at different taxonomic levels.</span></span>
<span><span class="co"># Specifically, we analyze enrichment at the family level using metagenomic data</span></span>
<span></span>
<span><span class="co"># Load required libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feargalr/taxsea" class="external-link">TaxSEA</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">curatedMetagenomicData</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">phyloseq</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">MicrobiomeStat</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load sample metadata</span></span>
<span><span class="va">metadata_all</span> <span class="op">&lt;-</span> <span class="va">sampleMetadata</span></span>
<span></span>
<span><span class="co"># Filter metadata for the specific study (ChngKR_2016)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="va">metadata_all</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">study_name</span> <span class="op">==</span> <span class="st">"ChngKR_2016"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">column_to_rownames</span><span class="op">(</span><span class="st">'sample_id'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract count data using curatedMetagenomicData</span></span>
<span><span class="va">cmd_data</span> <span class="op">&lt;-</span> <span class="fu">curatedMetagenomicData</span><span class="op">(</span></span>
<span>  pattern <span class="op">=</span> <span class="st">"ChngKR_2016.relative_abundance"</span>,</span>
<span>  counts <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  dryrun <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert the extracted data to a count matrix</span></span>
<span><span class="va">counts_data</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">cmd_data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">counts_data</span> <span class="op">&lt;-</span> <span class="va">counts_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span><span class="op">]</span>  <span class="co"># Subset to relevant samples</span></span>
<span></span>
<span><span class="co"># Filter taxa with at least one sample having counts &gt; 100</span></span>
<span><span class="va">counts_data</span> <span class="op">&lt;-</span> <span class="va">counts_data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">counts_data</span> <span class="op">&gt;</span> <span class="fl">100</span>, <span class="fl">1</span>, <span class="va">sum</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Extract species names from taxonomic strings</span></span>
<span><span class="va">species_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"s__"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts_data</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">y</span>, <span class="st">"\\|"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">7</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">species_names</span></span>
<span></span>
<span><span class="co"># Create a taxonomic lineage dataframe</span></span>
<span><span class="co"># Remove taxonomic prefixes (k__, p__, c__, etc.) and separate into taxonomic ranks</span></span>
<span></span>
<span><span class="co"># make data frame of taxon lineages</span></span>
<span><span class="va">taxon_lineages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Name <span class="op">=</span> <span class="va">species_names</span>,</span>
<span>                             Lineage <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">species_names</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Lineage <span class="op">=</span> <span class="fu">str_remove_all</span><span class="op">(</span><span class="va">Lineage</span>, <span class="st">'[kpcofgs]__'</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">separate</span><span class="op">(</span>col <span class="op">=</span> <span class="va">Lineage</span>, into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'kingdom'</span>, <span class="st">'phylum'</span>, <span class="st">'class'</span>, </span>
<span>                                   <span class="st">'order'</span>, <span class="st">'family'</span>, <span class="st">'genus'</span>, <span class="st">'species'</span><span class="op">)</span>, </span>
<span>           sep <span class="op">=</span> <span class="st">'\\|'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>name <span class="op">=</span> <span class="va">Name</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">remove_rownames</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">column_to_rownames</span><span class="op">(</span><span class="st">'name'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform differential abundance testing using LinDA</span></span>
<span><span class="va">metadata</span><span class="op">$</span><span class="va">study_condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">study_condition</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"AD"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">linda_results</span> <span class="op">&lt;-</span> <span class="fu">linda</span><span class="op">(</span></span>
<span>  feature.dat <span class="op">=</span> <span class="va">counts_data</span>,</span>
<span>  meta.dat <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>  formula <span class="op">=</span> <span class="st">'~study_condition'</span>,</span>
<span>  feature.dat.type <span class="op">=</span> <span class="st">'count'</span>,</span>
<span>  prev.filter <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract log2 fold change values for differential taxa</span></span>
<span><span class="va">linda_results</span> <span class="op">&lt;-</span> <span class="va">linda_results</span><span class="op">$</span><span class="va">output</span><span class="op">$</span><span class="va">study_conditionAD</span></span>
<span><span class="va">log2_fold_changes</span> <span class="op">&lt;-</span> <span class="va">linda_results</span><span class="op">$</span><span class="va">log2FoldChange</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">log2_fold_changes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">linda_results</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define the taxonomic rank for enrichment analysis</span></span>
<span><span class="va">selected_taxon_level</span> <span class="op">&lt;-</span> <span class="st">'genus'</span>  <span class="co"># Modify as needed (e.g., genus, phylum)</span></span>
<span></span>
<span><span class="co"># Create a named list of species grouped by taxonomic rank</span></span>
<span><span class="va">custom_taxon_sets</span> <span class="op">&lt;-</span> <span class="va">taxon_lineages</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">.data</span><span class="op">[[</span><span class="va">selected_taxon_level</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">species</span><span class="op">)</span>, .groups <span class="op">=</span> <span class="st">"drop"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">deframe</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform enrichment analysis using TaxSEA</span></span>
<span><span class="va">custom_taxsea_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TaxSEA.html">TaxSEA</a></span><span class="op">(</span>taxon_ranks <span class="op">=</span> <span class="va">log2_fold_changes</span>, custom_db <span class="op">=</span> <span class="va">custom_taxon_sets</span><span class="op">)</span></span>
<span><span class="va">custom_taxsea_results</span> <span class="op">&lt;-</span> <span class="va">custom_taxsea_results</span><span class="op">$</span><span class="va">custom_sets</span></span></code></pre></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Feargal Ryan.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
