<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>TaxSEA • TaxSEA</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="TaxSEA">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TaxSEA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-guides" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Guides</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-guides">
<li><a class="dropdown-item" href="../articles/why-taxon-set-enrichment.html">Why use taxon set enrichment?</a></li>
    <li><a class="dropdown-item" href="../articles/FAQs.html">FAQs</a></li>
    <li><a class="dropdown-item" href="../articles/Taxonomic-aggregation.html">Taxonomic aggregation</a></li>
    <li><a class="dropdown-item" href="../articles/TaxSEA.html">Using TaxSEA</a></li>
    <li><a class="dropdown-item" href="../articles/ORA_vs_ES.html">Analysis types</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/feargalr/taxsea/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>TaxSEA</h1>
                        <h4 data-toc-skip class="author">Feargal
Ryan</h4>
            
            <h4 data-toc-skip class="date">2026-01-15</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/feargalr/taxsea/blob/main/vignettes/TaxSEA.Rmd" class="external-link"><code>vignettes/TaxSEA.Rmd</code></a></small>
      <div class="d-none name"><code>TaxSEA.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="taxsea-taxon-set-enrichment-analysis">TaxSEA: Taxon Set Enrichment Analysis<a class="anchor" aria-label="anchor" href="#taxsea-taxon-set-enrichment-analysis"></a>
</h2>
<p>TaxSEA is an R package designed to enable rapid interpretation of
differential abundance analysis or correlation analysis output for
microbiota data. TaxSEA takes as input a vector of genus or species
names and a rank. For example log2 fold changes, or Spearman’s rho.
TaxSEA then uses a Kolmogorov-Smirnov test to identify if a particular
group of species or genera (i.e. a set of taxa such as butyrate
producers) are skewed to one end of the distribution .</p>
<p>Simply put, TaxSEA allows users to rapidly go from a list of species
to which metabolite producers are altered, and if the findings are
similar to a previously published study.</p>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>To install the latest version of TaxSEA from Bioconductor:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"TaxSEA"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="taxon-set-database">Taxon set database<a class="anchor" aria-label="anchor" href="#taxon-set-database"></a>
</h3>
<p>TaxSEA utilizes taxon sets generated from five reference databases
(<strong>gutMGene</strong>, <strong>GMrepo v2</strong>,
<strong>MiMeDB</strong>, <strong>mBodyMap</strong>,
<strong>BugSigDB</strong>).</p>
<p>Please cite the appropriate database if using:</p>
<ul>
<li><p>Cheng et al. gutMGene: a comprehensive database for target genes
of gut microbes and microbial metabolites Nucleic Acids Res.
2022.</p></li>
<li><p>Dai et al. GMrepo v2: a curated human gut microbiome database
with special focus on disease markers and cross-dataset comparison
Nucleic Acids Res. 2022.</p></li>
<li><p>Wishart et. al. MiMeDB: the Human Microbial Metabolome Database
Nucleic Acids Res. 2023.</p></li>
<li><p>Jin et al. mBodyMap: a curated database for microbes across human
body and their associations with health and diseases. Nucleic Acids Res.
2022.</p></li>
<li><p>Geistlinger et al. BugSigDB captures patterns of differential
abundance across a broad range of host-associated microbial signatures.
Nature Biotech. 2023.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="test-data">Test data<a class="anchor" aria-label="anchor" href="#test-data"></a>
</h3>
<p>The test data provided with TaxSEA consists of log2 fold changes
comparing between healthy and IBD. The count data for this was
downloaded from curatedMetagenomeData and fold changes generated with
LinDA.</p>
<ul>
<li>Hall et al. A novel Ruminococcus gnavus clade enriched in
inflammatory bowel disease patients Genome Med. 2017 Nov
28;9(1):103.</li>
<li>Pasolli et al. Accessible, curated metagenomic data through
ExperimentHub. Nat Methods. 2017 Oct 31;14(11):1023-1024. doi:
10.1038/nmeth.4468.</li>
<li>Zhou et al. LinDA: linear models for differential abundance analysis
of microbiome compositional data. Genome Biol. 2022 Apr
14;23(1):95.</li>
</ul>
</div>
<div class="section level3">
<h3 id="functions">Functions<a class="anchor" aria-label="anchor" href="#functions"></a>
</h3>
<ul>
<li>
<code>get_taxon_sets(taxon)</code>: Retrieves taxon sets which
contain a particular taxon for a list of taxon names.</li>
<li>
<code>get_ncbi_taxon_ids(taxon_names)</code>: Retrieves NCBI
Taxonomy IDs for a list of taxon names.</li>
<li>
<code>TaxSEA(taxon_ranks, database = "All")</code>: Taxon set
enrichment analysis.</li>
</ul>
</div>
<div class="section level3">
<h3 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h3>
<div class="section level4">
<h4 id="retrieve-sets-containing-a-particular-taxon">Retrieve sets containing a particular taxon<a class="anchor" aria-label="anchor" href="#retrieve-sets-containing-a-particular-taxon"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feargalr/taxsea" class="external-link">TaxSEA</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Retrieve taxon sets containing Bifidobacterium longum.</span></span>
<span><span class="va">blong.sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_taxon_sets.html">get_taxon_sets</a></span><span class="op">(</span>taxon<span class="op">=</span><span class="st">"Bifidobacterium_longum"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="running-an-enrichment-analysis">Running an enrichment analysis<a class="anchor" aria-label="anchor" href="#running-an-enrichment-analysis"></a>
</h4>
<p>All that is required for TaxSEA is a named vector of log2 fold
changes between groups for species or genera. TaxSEA will not work for
ranks higher than species or genus. The input should be for all taxa
tested, and not limited to only a pre-defined set (e.g. do not use a
threshold for significance or remove any taxa). See example below for
format. TaxSEA will lookup and convert taxon names to NCBI taxonomic
identifiers. TaxSEA stores commonly observed identifiers internally to
save time.</p>
<p>TaxSEA can also utilise custom databases which should be a named list
of taxon sets. In this case the ID conversion is disabled and it is
expected that the input names and database names will be in the same
format</p>
<p>Input IDs should be in the format of like one of the following -
Species name. E.g. “Bifidobacterium longum”, “Bifidobacterium_longum” -
Genus name. E.g. “Bifidobacterium” - NCBI ID E.g. 216816</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Input IDs with the full taxonomic lineage should be split up. E.g.</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>  <span class="st">"d__Bacteria.p__Actinobacteriota.c__Actinomycetes."</span>,</span>
<span>  <span class="st">"o__Bifidobacteriales.f__Bifidobacteriaceae.g__Bifidobacterium"</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">x</span>,split<span class="op">=</span><span class="st">"\\."</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"g__"</span>,<span class="st">""</span>,<span class="va">x</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Bifidobacterium"</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Example test data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feargalr/taxsea" class="external-link">TaxSEA</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">TaxSEA_test_data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">TaxSEA_test_data</span><span class="op">)</span>,<span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   Holdemania_filiformis  Clostridium_sp_CAG_242 Bifidobacterium_bifidum </span></span>
<span><span class="co">##                  -1.609                   0.275                   3.146 </span></span>
<span><span class="co">##        Actinomyces_oris </span></span>
<span><span class="co">##                   1.336</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="run-taxsea-with-test-data">Run TaxSEA with test data<a class="anchor" aria-label="anchor" href="#run-taxsea-with-test-data"></a>
</h4>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"TaxSEA_test_data"</span><span class="op">)</span></span>
<span><span class="va">taxsea_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TaxSEA.html">TaxSEA</a></span><span class="op">(</span>taxon_ranks<span class="op">=</span><span class="va">TaxSEA_test_data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Using cached version from 2026-01-15 00:22:00</span></span></code></pre>
<pre><code><span><span class="co">## Warning in ks.test.default(taxon_set_ranks, taxon_ranks): p-value will be</span></span>
<span><span class="co">## approximate in the presence of ties</span></span>
<span><span class="co">## Warning in ks.test.default(taxon_set_ranks, taxon_ranks): p-value will be</span></span>
<span><span class="co">## approximate in the presence of ties</span></span>
<span><span class="co">## Warning in ks.test.default(taxon_set_ranks, taxon_ranks): p-value will be</span></span>
<span><span class="co">## approximate in the presence of ties</span></span>
<span><span class="co">## Warning in ks.test.default(taxon_set_ranks, taxon_ranks): p-value will be</span></span>
<span><span class="co">## approximate in the presence of ties</span></span>
<span><span class="co">## Warning in ks.test.default(taxon_set_ranks, taxon_ranks): p-value will be</span></span>
<span><span class="co">## approximate in the presence of ties</span></span>
<span><span class="co">## Warning in ks.test.default(taxon_set_ranks, taxon_ranks): p-value will be</span></span>
<span><span class="co">## approximate in the presence of ties</span></span>
<span><span class="co">## Warning in ks.test.default(taxon_set_ranks, taxon_ranks): p-value will be</span></span>
<span><span class="co">## approximate in the presence of ties</span></span>
<span><span class="co">## Warning in ks.test.default(taxon_set_ranks, taxon_ranks): p-value will be</span></span>
<span><span class="co">## approximate in the presence of ties</span></span>
<span><span class="co">## Warning in ks.test.default(taxon_set_ranks, taxon_ranks): p-value will be</span></span>
<span><span class="co">## approximate in the presence of ties</span></span>
<span><span class="co">## Warning in ks.test.default(taxon_set_ranks, taxon_ranks): p-value will be</span></span>
<span><span class="co">## approximate in the presence of ties</span></span>
<span><span class="co">## Warning in ks.test.default(taxon_set_ranks, taxon_ranks): p-value will be</span></span>
<span><span class="co">## approximate in the presence of ties</span></span>
<span><span class="co">## Warning in ks.test.default(taxon_set_ranks, taxon_ranks): p-value will be</span></span>
<span><span class="co">## approximate in the presence of ties</span></span>
<span><span class="co">## Warning in ks.test.default(taxon_set_ranks, taxon_ranks): p-value will be</span></span>
<span><span class="co">## approximate in the presence of ties</span></span>
<span><span class="co">## Warning in ks.test.default(taxon_set_ranks, taxon_ranks): p-value will be</span></span>
<span><span class="co">## approximate in the presence of ties</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Enrichments among metabolite producers from gutMgene and MiMeDB</span></span>
<span><span class="va">metabolites.df</span> <span class="op">=</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">Metabolite_producers</span></span>
<span></span>
<span><span class="co">#Enrichments among health and disease signatures from GMRepoV2 and mBodyMap</span></span>
<span><span class="va">disease.df</span> <span class="op">=</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">Health_associations</span></span>
<span></span>
<span><span class="co">#Enrichments amongh published associations from BugSigDB</span></span>
<span><span class="va">bsdb.df</span> <span class="op">=</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">BugSigdB</span></span></code></pre></div>
<div class="section level5">
<h5 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h5>
<p>The output is a list of three dataframes providing enrichment results
for metabolite produers, health/disease associations, and published
signatures from BugSigDB. Each dataframe has 5 columns - taxonSetName -
The name of the taxon set tested - median_rank - The median rank of set
members - P value - Kolmogorov-Smirnov test P value. - FDR - P value
adjusted for multiple testing. - TaxonSet - Returns list of taxa in the
set to show what is driving the signal</p>
</div>
<div class="section level5">
<h5 id="bugsigdb">BugSigDB<a class="anchor" aria-label="anchor" href="#bugsigdb"></a>
</h5>
<p>The format of BugSigDB is that each publication is entered as a
“Study”, and within this there is different experiments and signatures.
Should users wish to find out more information about the signatures,
they can do so by querying that database.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/waldronlab/bugsigdbr" class="external-link">bugsigdbr</a></span><span class="op">)</span> <span class="co">#This package is installable via Bioconductor</span></span>
<span><span class="va">bsdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bugsigdbr/man/importBugSigDB.html" class="external-link">importBugSigDB</a></span><span class="op">(</span><span class="op">)</span> <span class="co">#Import database </span></span></code></pre></div>
<pre><code><span><span class="co">## Using cached version from 2026-01-15 00:22:00</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#E.g. if the BugSigDB identifier you found enriched was </span></span>
<span><span class="co">#bsdb:74/1/2_obesity:obese_vs_non-obese_DOWN</span></span>
<span><span class="co">#This is Study 74, Experiment 1, Signature 2</span></span>
<span><span class="va">bsdb</span><span class="op">[</span><span class="va">bsdb</span><span class="op">$</span><span class="va">Study</span><span class="op">==</span><span class="st">"Study 74"</span> <span class="op">&amp;</span> </span>
<span>     <span class="va">bsdb</span><span class="op">$</span><span class="va">Experiment</span><span class="op">==</span><span class="st">"Experiment 1"</span> <span class="op">&amp;</span> </span>
<span>     <span class="va">bsdb</span><span class="op">$</span><span class="va">Signature</span><span class="op">==</span><span class="st">"Signature 2"</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] BSDB ID                    Study                     </span></span>
<span><span class="co">##  [3] Study design               PMID                      </span></span>
<span><span class="co">##  [5] DOI                        URL                       </span></span>
<span><span class="co">##  [7] Authors list               Title                     </span></span>
<span><span class="co">##  [9] Journal                    Year                      </span></span>
<span><span class="co">## [11] Keywords                   Experiment                </span></span>
<span><span class="co">## [13] Location of subjects       Host species              </span></span>
<span><span class="co">## [15] Body site                  UBERON ID                 </span></span>
<span><span class="co">## [17] Condition                  EFO ID                    </span></span>
<span><span class="co">## [19] Group 0 name               Group 1 name              </span></span>
<span><span class="co">## [21] Group 1 definition         Group 0 sample size       </span></span>
<span><span class="co">## [23] Group 1 sample size        Antibiotics exclusion     </span></span>
<span><span class="co">## [25] Sequencing type            16S variable region       </span></span>
<span><span class="co">## [27] Sequencing platform        Statistical test          </span></span>
<span><span class="co">## [29] Significance threshold     MHT correction            </span></span>
<span><span class="co">## [31] LDA Score above            Matched on                </span></span>
<span><span class="co">## [33] Confounders controlled for Pielou                    </span></span>
<span><span class="co">## [35] Shannon                    Chao1                     </span></span>
<span><span class="co">## [37] Simpson                    Inverse Simpson           </span></span>
<span><span class="co">## [39] Richness                   Signature page name       </span></span>
<span><span class="co">## [41] Source                     Curated date              </span></span>
<span><span class="co">## [43] Curator                    Revision editor           </span></span>
<span><span class="co">## [45] Description                Abundance in Group 1      </span></span>
<span><span class="co">## [47] MetaPhlAn taxon names      NCBI Taxonomy IDs         </span></span>
<span><span class="co">## [49] State                      Reviewer                  </span></span>
<span><span class="co">## &lt;0 rows&gt; (or 0-length row.names)</span></span></code></pre>
</div>
<div class="section level5">
<h5 id="taxsea-database-with-other-enrichment-tools">TaxSEA database with other enrichment tools<a class="anchor" aria-label="anchor" href="#taxsea-database-with-other-enrichment-tools"></a>
</h5>
<p>The TaxSEA function by default uses the Kolmogorov Smirnov test and
the original idea was inspired by gene set enrichment analyses from
RNASeq. Should users wish to use an alternative gene set enrichment
analysis tool the database is formatted in such a way that should be
possible. See below for an example with fast gene set enrichment
analysis (fgsea).</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ctlab/fgsea/" class="external-link">fgsea</a></span><span class="op">)</span> <span class="co">#This package is installable via Bioconductor</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"TaxSEA_test_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"TaxSEA_db"</span><span class="op">)</span></span>
<span><span class="co">#Convert input names to NCBI taxon ids</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">TaxSEA_test_data</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_ncbi_taxon_ids.html">get_ncbi_taxon_ids</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">TaxSEA_test_data</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">TaxSEA_test_data</span> <span class="op">=</span> <span class="va">TaxSEA_test_data</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">TaxSEA_test_data</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#Run fgsea</span></span>
<span><span class="va">fgsea_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/fgsea/man/fgsea.html" class="external-link">fgsea</a></span><span class="op">(</span><span class="va">TaxSEA_db</span>, <span class="va">TaxSEA_test_data</span>, minSize<span class="op">=</span><span class="fl">5</span>, maxSize<span class="op">=</span><span class="fl">500</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, : There are ties in the preranked stats (1.83% of the list).</span></span>
<span><span class="co">## The order of those tied genes will be arbitrary, which may produce unexpected results.</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] fgsea_1.36.2     bugsigdbr_1.16.2 TaxSEA_1.3.1     BiocStyle_2.38.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] fastmatch_1.1-6     gtable_0.3.6        xfun_0.55          </span></span>
<span><span class="co">##  [4] bslib_0.9.0         ggplot2_4.0.1       httr2_1.2.2        </span></span>
<span><span class="co">##  [7] lattice_0.22-7      vctrs_0.6.5         tools_4.5.2        </span></span>
<span><span class="co">## [10] generics_0.1.4      curl_7.0.0          parallel_4.5.2     </span></span>
<span><span class="co">## [13] tibble_3.3.1        RSQLite_2.4.5       blob_1.2.4         </span></span>
<span><span class="co">## [16] pkgconfig_2.0.3     Matrix_1.7-4        data.table_1.18.0  </span></span>
<span><span class="co">## [19] dbplyr_2.5.1        RColorBrewer_1.1-3  S7_0.2.1           </span></span>
<span><span class="co">## [22] desc_1.4.3          lifecycle_1.0.5     compiler_4.5.2     </span></span>
<span><span class="co">## [25] farver_2.1.2        textshaping_1.0.4   codetools_0.2-20   </span></span>
<span><span class="co">## [28] htmltools_0.5.9     sass_0.4.10         yaml_2.3.12        </span></span>
<span><span class="co">## [31] pillar_1.11.1       pkgdown_2.2.0       jquerylib_0.1.4    </span></span>
<span><span class="co">## [34] BiocParallel_1.44.0 cachem_1.1.0        tidyselect_1.2.1   </span></span>
<span><span class="co">## [37] digest_0.6.39       dplyr_1.1.4         purrr_1.2.1        </span></span>
<span><span class="co">## [40] bookdown_0.46       cowplot_1.2.0       fastmap_1.2.0      </span></span>
<span><span class="co">## [43] grid_4.5.2          cli_3.6.5           magrittr_2.0.4     </span></span>
<span><span class="co">## [46] withr_3.0.2         filelock_1.0.3      scales_1.4.0       </span></span>
<span><span class="co">## [49] rappdirs_0.3.3      bit64_4.6.0-1       rmarkdown_2.30     </span></span>
<span><span class="co">## [52] bit_4.6.0           ragg_1.5.0          memoise_2.0.1      </span></span>
<span><span class="co">## [55] evaluate_1.0.5      knitr_1.51          BiocFileCache_3.0.0</span></span>
<span><span class="co">## [58] rlang_1.1.7         Rcpp_1.1.1          glue_1.8.0         </span></span>
<span><span class="co">## [61] DBI_1.2.3           BiocManager_1.30.27 jsonlite_2.0.0     </span></span>
<span><span class="co">## [64] R6_2.6.1            systemfonts_1.3.1   fs_1.6.6</span></span></code></pre>
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Feargal Ryan.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
